### PR TITLE
[CALCITE-6255] Support BigQuery-style JSON_OBJECT invocation syntax

### PR DESCRIPTION
**Pull request by barrkel**
[CALCITE-6255](https://issues.apache.org/jira/browse/CALCITE-6255)

Currently, Calcite has two styles for invoking `JSON_OBJECT`:

```
    JSON_OBJECT(KEY 'foo' VALUE 'bar', KEY 'foo2' VALUE 'bar2')
```

and:

```
    JSON_OBJECT('foo' : 'bar', 'foo2' : 'bar')
```

However, [BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/json_functions#json_object) uses a lighter-weight syntax resembling the MAP<> constructor, with an even numbered argument list where every odd item is a key and the following even item is a value:

```
    JSON_OBJECT('foo', 'bar', 'foo2', 'bar2')
```

This PR updates the grammar to allow `,` as well as `:` for separating key/value pairs in the invocation.

It does not prevent mixing `,` and `:`, but retains the prohibition from mixing with `KEY`/`VALUE` syntax.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java [hunk 1]
```diff
@@ -112,6 +112,9 @@ public interface CalciteResource {
   @BaseMessage("Illegal identifier '':''. Was expecting ''VALUE''")
   ExInst<CalciteException> illegalColon();
 
+  @BaseMessage("Illegal identifier '',''. Was expecting ''VALUE''")
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/runtime/CalciteResource.java:115]: Comma is not really an identifier.  Maybe "symbol".
Yes, the previous message is probably wrong too.

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java [hunk 2]
```diff
@@ -8815,6 +8815,10 @@ private static Consumer<List<? extends Throwable>> checkWarnings(
         .ok("JSON_OBJECT(KEY 'foo' VALUE "
             + "JSON_OBJECT(KEY 'foo' VALUE 'bar' NULL ON NULL) "
             + "FORMAT JSON NULL ON NULL)");
+    expr("json_object('foo', 'bar')")
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:8818]: how about some other tests, e.g., with an odd number of elements?

File: core/src/main/codegen/templates/Parser.jj [hunk 3]
```diff
@@ -6767,7 +6773,8 @@ SqlCall JsonObjectFunctionCall() :
 }
 {
     <JSON_OBJECT> { span = span(); }
-    <LPAREN> [
+    <LPAREN>
```
Comment (mihaibudiu) [inline core/src/main/codegen/templates/Parser.jj:6776]: why this change?

**Review by tanclary — COMMENTED** [review]
format the PR title and commit so that they begin with `[CALCITE-6255]` 

otherwise LGTM

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java (hunk 1 above)
Comment (tanclary) [reply core/src/main/java/org/apache/calcite/runtime/CalciteResource.java:115]: It's kind of a delimiter too? Wdyt

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/runtime/CalciteResource.java:115]: delimiter is also fine

File: testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java (hunk 2 above)
Comment (JiajunBernoulli) [reply testkit/src/main/java/org/apache/calcite/sql/parser/SqlParserTest.java:8818]: How about mixed syntax tests? e.g.
```
json_object('foo': json_object('foo', 'bar') format json)
```

File: core/src/main/codegen/templates/Parser.jj (hunk 3 above)
Comment (barrkel) [reply core/src/main/codegen/templates/Parser.jj:6776]: Stray edit from an earlier version of the code. I'll revert.

File: core/src/main/java/org/apache/calcite/runtime/CalciteResource.java (hunk 1 above)
Comment (barrkel) [reply core/src/main/java/org/apache/calcite/runtime/CalciteResource.java:115]: Well `,` isn't a legal identifier - the message is not wrong. It's kind of a two-prong message; we were expecting an identifier, and for the identifier to be `VALUE`. We fell at the first hurdle because `,` isn't an identifier.

(AFAICT `VALUE` is usually a keyword not an identifier but it's a bit squishy because it could easily be a non-reserved word.)

A better message might be "Unexpected symbol" rather than "Illegal identifier", which has better symmetry with the expected clause following. I'll change both to that.

**Review by tanclary — APPROVED** [review]
LGTM, will merge soon if no other comments.

**Review by macroguo-ghy — APPROVED** [review]
+1

**Comment by macroguo-ghy** [comment]
I think you should make the Jira ticket title, GitHub PR title, and commit message title consistent.

**Comment by barrkel** [comment]
> I think you should make the Jira ticket title, GitHub PR title, and commit message title consistent.

Done.

**Review by tanclary — APPROVED** [review]
LGTM
